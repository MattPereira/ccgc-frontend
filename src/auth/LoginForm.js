import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import {
  Card,
  CardBody,
  CardTitle,
  Form,
  Label,
  Input,
  FormGroup,
  Alert,
} from "reactstrap";

/** Login form.
 *
 * Shows form and manages updates to state on changes.
 * On submission:
 * - calls login function prop
 * - redirects to "/" route
 *
 * Routes -> SignupForm
 * Routed as /login
 */

const LoginForm = ({ login }) => {
  const navigate = useNavigate();
  const [formData, setFormData] = useState({
    email: "",
    password: "",
  });

  const [formErrors, setFormErrors] = useState([]);

  console.debug(
    "LoginForm",
    "login=",
    typeof login,
    "formData=",
    formData,
    "formErrors=",
    formErrors
  );

  //update state of formData onChange of any form input field
  const handleChange = (e) => {
    const { name, value } = e.target;

    setFormData((fData) => ({
      ...fData,
      [name]: value,
    }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    let result = await login(formData);
    if (result.success) {
      navigate("/");
    } else {
      setFormErrors(result.errors);
    }
  };

  return (
    <div className="row justify-content-center">
      <div className="col-sm-8">
        <Card className="px-5 py-3">
          <CardBody>
            <CardTitle className="display-4 text-center">Login</CardTitle>

            <Form onSubmit={handleSubmit}>
              <FormGroup>
                <Label htmlFor="email">email</Label>
                <Input
                  className="form-control"
                  id="email"
                  name="email"
                  type="text"
                  onChange={handleChange}
                  required
                  autoComplete="email"
                ></Input>
              </FormGroup>
              <FormGroup>
                <Label htmlFor="password">Password</Label>
                <Input
                  className="form-control"
                  id="password"
                  name="password"
                  type="password"
                  onChange={handleChange}
                  required
                  autoComplete="current-password"
                ></Input>
              </FormGroup>

              {formErrors.length
                ? formErrors.map((err) => (
                    <Alert color="danger" key={err}>
                      {err}
                    </Alert>
                  ))
                : null}

              <div className="row justify-content-end">
                <div className="col-auto">
                  <button className="btn btn-primary btn-block px-4">
                    submit
                  </button>
                </div>
              </div>
            </Form>
          </CardBody>
        </Card>
      </div>
    </div>
  );
};

export default LoginForm;
