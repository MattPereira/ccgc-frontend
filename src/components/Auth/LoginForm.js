import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import { Card, Form, Alert, Container } from "react-bootstrap";
import { Button } from "@mui/material";

/** Login form.
 *
 * Shows form and manages updates to state on changes.
 * On submission:
 * - calls login function prop
 * - redirects to "/" route
 *
 * Router -> RegisterForm
 * Routed to "/login"
 */

const LoginForm = ({ login }) => {
  const navigate = useNavigate();
  const [formData, setFormData] = useState({
    email: "",
    password: "",
  });

  const [formErrors, setFormErrors] = useState([]);

  console.debug(
    "LoginForm",
    "login=",
    typeof login,
    "formData=",
    formData,
    "formErrors=",
    formErrors
  );

  //update state of formData onChange of any form input field
  const handleChange = (e) => {
    const { name, value } = e.target;

    setFormData((fData) => ({
      ...fData,
      [name]: value,
    }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();

    formData.email = formData.email.toLowerCase();
    console.log(formData.email);
    let result = await login(formData);
    if (result.success) {
      navigate("/");
    } else {
      setFormErrors(result.errors);
    }
  };

  return (
    <Container className="py-5">
      <div className="row justify-content-center">
        <div className="col-sm-10 col-md-6">
          <h1 className="text-center display-1 mb-5">Login</h1>

          <Card className="shadow mb-5">
            <Card.Img
              src="https://images.unsplash.com/photo-1623567341691-1f47b5cf949e?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1752&q=80"
              alt="Login card image"
            />

            <Card.Body className="px-5 py-5">
              <Form onSubmit={handleSubmit}>
                <Form.Group className="mb-3">
                  <Form.Label htmlFor="email">Email</Form.Label>
                  <input
                    className="form-control"
                    id="email"
                    name="email"
                    type="text"
                    onChange={handleChange}
                    required
                    autoComplete="email"
                  ></input>
                </Form.Group>
                <Form.Group className="mb-3">
                  <Form.Label htmlFor="password">Password</Form.Label>
                  <input
                    className="form-control"
                    id="password"
                    name="password"
                    type="password"
                    onChange={handleChange}
                    required
                    autoComplete="current-password"
                  ></input>
                </Form.Group>

                {formErrors.length
                  ? formErrors.map((err) => (
                      <Alert variant="danger" key={err}>
                        {err}
                      </Alert>
                    ))
                  : null}

                <div className="text-end">
                  <Button variant="contained" type="submit">
                    Login
                  </Button>
                </div>
              </Form>
            </Card.Body>
          </Card>
        </div>
      </div>
    </Container>
  );
};

export default LoginForm;
