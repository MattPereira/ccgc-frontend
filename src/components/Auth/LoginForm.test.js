import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import LoginForm from "./LoginForm";

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <LoginForm />
    </MemoryRouter>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <LoginForm />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("displays form inputs and login button", function () {
  render(
    <MemoryRouter initialEntries={["/login"]}>
      <LoginForm />
    </MemoryRouter>
  );

  const emailInput = screen.getByLabelText("Email");
  expect(emailInput).toBeInTheDocument();

  const passwordInput = screen.getByLabelText("Password");
  expect(passwordInput).toBeInTheDocument();

  //only one button in the component
  const loginBtn = screen.getByRole("button");
  expect(loginBtn).toBeInTheDocument();
});
