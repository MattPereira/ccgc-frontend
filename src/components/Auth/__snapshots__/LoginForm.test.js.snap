// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="row justify-content-center"
  >
    <div
      class="col-sm-10 col-md-8"
    >
      <div
        class="shadow mb-5 card"
      >
        <div
          class="display-2 text-center bg-dark text-white py-2 mb-0 card-title h5"
        >
          Login
        </div>
        <img
          alt="Login card image"
          class="card-img"
          src="https://images.unsplash.com/photo-1623567341691-1f47b5cf949e?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1752&q=80"
        />
        <div
          class="px-5 py-5 card-body"
        >
          <form
            class=""
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="email"
              >
                Email
              </label>
              <input
                autocomplete="email"
                class="form-control"
                id="email"
                name="email"
                required=""
                type="text"
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="password"
              >
                Password
              </label>
              <input
                autocomplete="current-password"
                class="form-control"
                id="password"
                name="password"
                required=""
                type="password"
              />
            </div>
            <div
              class="row justify-content-end"
            >
              <div
                class="col-auto"
              >
                <button
                  class="btn btn-primary btn-block px-4"
                >
                  Login
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
